### ISSUE TITLE
discovery: clarifications on use of multiple meta tags

### ISSUE THREAD

**Issue by sgotti** [appc/spec#117]
What's the pourpose of multiple discovery meta tags?

1. Support different aci names formats:

Example (see also #61):

I'm imagining that in my repository I can have also some "noarch" packages.
For getting this, using meta discovery, I should provide two meta tags like these:

```
<meta name="ac-discovery" content="example.com https://storage.example.com/{name}-{version}-{os}-{arch}.{ext}">
<meta name="ac-discovery" content="example.com https://storage.example.com/{name}-{version}.{ext}">
```

Trying to discover an aci without providing os and arch, there can be two options:
* If the ACE, on discovery, automatically sets os and arch to default values (like rocket does), the first template should fail download as there isn't an ACI with that name and the second will work.
* If the ACE, on discovery, doesn't set default values for os and arch, the first template will fail to render and the second will succeed and will be downloaded.

Both gives the expected result to get "noarch" acis.

A possible downside is that the user can get various warning as some downloads will "correctly" fail. 

2. Multiple mirrors.

```
<meta name="ac-discovery" content="example.com https://storage.example.com/{name}-{version}-{os}-{arch}.{ext}">
<meta name="ac-discovery" content="example.com https://mirror.storage.example.com/{name}-{version}-{os}-{arch}.{ext}">
```

For this to fully work there should be a mirror choosing logic done from the ACE during fetching.

3. Both 1 and 2.

```
<meta name="ac-discovery" content="example.com https://storage.example.com/{name}-{version}-{os}-{arch}.{ext}">
<meta name="ac-discovery" content="example.com https://mirror.storage.example.com/{name}-{version}-{os}-{arch}.{ext}">
<meta name="ac-discovery" content="example.com https://storage.example.com/{name}-{version}.{ext}">
<meta name="ac-discovery" content="example.com https://mirror.storage.example.com/{name}-{version}.{ext}">
```

I think that all these use cases should work with the above downsides:
* various warning as some downloads will "correctly" fail.

Do you see other problems or better solutions for the various points?

If so, do you think that the spec should be enhanced with some examples to clarify the various use cases.

**Comment by jonboulle** [comment]
> Multiple mirrors.

Quick drive-by comment: one of the key purposes is to support not just multiple (HTTP) mirrors but multiple _protocols_. The classic example would be an image that is retrievable over either HTTP or BitTorrent (actually I think we use HDFS as an example in the spec) - and the client doing the discovery can choose whichever it prefers.

**Comment by sgotti** [comment]
@jonboulle Right! I forgot about it (sorry).
Do you think that this purpose is going to clash with "multiple mirrors" and "different aci names formats"?
Do you see other problems or better solutions for the various points?

Thanks!

**Comment by philips** [comment]
Hrm, I think that if you are going to have a noarch application and an arch application it might be easier to make the prefixes different. Such as:

```
<meta name="ac-discovery" content="example.com/app https://storage.example.com/{name}-{version}-{os}-{arch}.{ext}">
<meta name="ac-discovery" content="example.com/app https://mirror.storage.example.com/{name}-{version}-{os}-{arch}.{ext}">
<meta name="ac-discovery" content="example.com/script https://storage.example.com/{name}-{version}.{ext}">
<meta name="ac-discovery" content="example.com/script https://mirror.storage.example.com/{name}-{version}.{ext}">
```

Perhaps I am missing something?

**Comment by sgotti** [comment]
@philips I didn't thought about the prefix match! It looks like a good idea.

This will increase the number of meta tags as there can be 1*Nmirrors ac-discovery meta tags for every application name. Or am I missing something?

Another possibility to reduce the number of metas will be to use different paths for noarch and arch apps (or make all arch apps to start with the same prefix, but it will be strange...). Something like `example.com/arch/` for arch apps and `example.com/noarch/` for no arch apps.

**Comment by jonboulle** [comment]
I believe this is addressed by #288 - @sgotti can you take a look?

**Comment by sgotti** [comment]
@jonboulle Sorry for the delay. #288 clarifies this. In a future (perhaps it'll be better to do this after #140) I'd like to sum up this issue in some discovery best practice examples. Thanks!
